### PR TITLE
Add stateful example for tutorial3

### PR DESCRIPTION
**Pull request by amityadav0**

### CODE REVIEW

File: examples/stateful-counter/README.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
+## Example
+
+We will create a simple stateful smart contract which will:
+- Have a global variable `counter`
+- Will increment `counter` each time we call the application.
+
+## Steps
+
++ Create a file with Approval Program
```
Comment (ratik21) [inline examples/stateful-counter/README.md:9]:
```suggestion
+ [Create a file with Approval Program](https://github.com/scale-it/algo-builder/blob/1307db9b620d8715c844bdf5c329599e557aade8/examples/stateful-counter/README.md#1-create-a-file-with-approval-program)
```
Can we hyperlink these points to their description below as well (as suggested above) ? note: probably the link is not right

File: examples/stateful-counter/README.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
++ Create a file with Clear program
++ Deploy New Application
++ Opt-In to Application
++ Call Application
++ Update Application
++ Delete Application
+
+# 1. Create a file with Approval Program
+
+We will create a approval program which has global `counter` and it is incremented when we call the application. Put the code below in `/assets/approval_program.teal`:
```
Comment (ratik21) [inline examples/stateful-counter/README.md:19]:
```suggestion
We will create an approval program which has global `counter` and it is incremented by 1 each time we call the application. Put the code below in `/assets/approval_program.teal`:
```

File: examples/stateful-counter/README.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+```javascript
+  const masterAccount = deployer.accountsByName.get('master-account');
+  const creatorAccount = deployer.accountsByName.get('alice');
+
+  const algoTxnParams = {
+    type: types.TransactionType.TransferAlgo,
+    sign: types.SignType.SecretKey,
+    fromAccount: masterAccount,
+    toAccountAddr: creatorAccount.addr,
+    amountMicroAlgos: 200000000,
```
Comment (ratik21) [inline examples/stateful-counter/README.md:75]:
```suggestion
    amountMicroAlgos: 200e6,
```

File: examples/stateful-counter/README.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    payFlags: {}
+  };
+  // transfer some algos to creator account
+  await executeTransaction(deployer, algoTxnParams);
+```
+
+In the code above we declare user accounts and fund `creatorAccount` account. `masterAccount` is the default account used in algob private net.
```
Comment (ratik21) [inline examples/stateful-counter/README.md:82]:
```suggestion
In the code above we declared user accounts and funded `creatorAccount` account. `masterAccount` is the default account used in algob private net.
```

File: examples/stateful-counter/README.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+### Deploy stateful contract
+
+Firstly we need to put some money into the contract. `master` account will fund it:
```
Comment (ratik21) [inline examples/stateful-counter/README.md:86]:
```suggestion
Firstly we need to fund the contract. `master` account will fund it (as it has lot of algos):
```

File: examples/stateful-counter/README.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    }, {});
+
+  console.log(sscInfo);
+```
+
+Parameters passed are:
+  - Approval Program
+  - Clear Program
+  - SSC(stateful smart contract) arguments which includes sender and details for storage usage.
+  - Transaction parameters
```
Comment (ratik21) [inline examples/stateful-counter/README.md:109]:
```suggestion
  - Common transaction parameters (eg. totalFee)
```

File: examples/stateful-counter/README.md [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+- In above code we have deployed a new application.
+
+- After deploying application you will get the following information on your terminal:
```
Comment (ratik21) [inline examples/stateful-counter/README.md:113]:
```suggestion
- After deploying application you should see the following information on your terminal:
```

File: examples/stateful-counter/README.md [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+```
+
+# 4. Opt-In to Application
+
+To Opt-In to an application use the following code in one of your scripts (in `./scripts`):
+
+```javascript
+	await deployer.optInToSSC(Account, applicationID, {}, {});
```
Comment (ratik21) [inline examples/stateful-counter/README.md:131]:
```suggestion
	await deployer.optInToSSC(account, applicationID, {}, {});
```

File: examples/stateful-counter/README.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+```
+
+where `Account` is the account you want to opt-in and applicationID is application index.
+
+# 5. Call Application
+
+To call an application use the following code in one of your scripts (in `./scripts`):
+
+```javascript
+	const tx = {
```
Comment (ratik21) [inline examples/stateful-counter/README.md:141]: Is the indentation here correct? Could you please check?

File: examples/stateful-counter/README.md [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+      appId: applicationID,
+      payFlags: {}
+    }
+
+    await executeTransaction(deployer, tx);
+```
+
+In `tx` there are following parameters:
+  - we set the type which is `CallNoOpSSC` - Call to stateful smart contract
+  - set the sign to SecretKey
```
Comment (ratik21) [inline examples/stateful-counter/README.md:154]:
```suggestion
  - set the sign to SecretKey (tx is signed by creatorAccount's sk)
```

File: examples/stateful-counter/README.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+  - provide fromAccount details
+  - provide application index of SSC
```
Comment (ratik21) [inline examples/stateful-counter/README.md:156]:
```suggestion
  - provide application index of SSC (retreived from checkpoint)
```

File: examples/stateful-counter/README.md [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+  - provide payment flags. If any (fee, firstValid, lastvalid etc)
```
Comment (ratik21) [inline examples/stateful-counter/README.md:157]:
```suggestion
  - provide payment flags, if any (eg. fee, firstValid, lastvalid etc)
```

File: examples/stateful-counter/README.md [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+```
+
+Output:
+
+```
+[ { key: 'Y291bnRlcg==', value: { bytes: '', type: 2, uint: 2 } } ]
+[ { key: 'Y291bnRlcg==', value: { bytes: '', type: 2, uint: 3 } } ]
+```
+
+here key 'Y291bnRlcg==' is converted form of `counter`.
```
Comment (ratik21) [inline examples/stateful-counter/README.md:175]:
```suggestion
here key 'Y291bnRlcg==' is base64 encoded form of `counter`.
```

File: examples/stateful-counter/README.md [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+
+# 6. Update Application
+
+To update application you can use:
```
Comment (ratik21) [inline examples/stateful-counter/README.md:179]:
```suggestion
To update an application with (new_approval.teal, new_clear.teal), you can use:
```

File: examples/stateful-counter/README.md [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+    'new_approval.teal',
+    'new_clear.teal',
+    {}
+  );
+console.log('Application Updated: ', updatedRes);
+```
+
+# 7. Delete Application
+
+To delete application you can use:
```
Comment (ratik21) [inline examples/stateful-counter/README.md:196]:
```suggestion
To delete an existing application you can use:
```

Comment (ratik21) [reply examples/stateful-counter/README.md:196]: We can also add a note that deleting non existing app will throw error

File: examples/stateful-counter/scripts/deploy.js [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+async function run (runtimeEnv, deployer) {
+  const masterAccount = deployer.accountsByName.get('master-account');
+  const creatorAccount = deployer.accountsByName.get('alice');
+
+  const algoTxnParams = {
+    type: types.TransactionType.TransferAlgo,
+    sign: types.SignType.SecretKey,
+    fromAccount: masterAccount,
+    toAccountAddr: creatorAccount.addr,
+    amountMicroAlgos: 200000000,
```
Comment (ratik21) [inline examples/stateful-counter/scripts/deploy.js:15]:
```suggestion
    amountMicroAlgos: 200e6,
```

File: examples/stateful-counter/README.md (hunk 1 above)
Comment (amityadav0) [reply examples/stateful-counter/README.md:9]: yes

File: examples/stateful-counter/README.md (hunk 9 above)
Comment (amityadav0) [reply examples/stateful-counter/README.md:141]: fixed

File: examples/stateful-counter/README.md [hunk 17]
```diff
@@ -0,0 +1,222 @@
+## Example
```
Comment (ratik21) [inline examples/stateful-counter/README.md:1]: Please rename the heading to something more meaningful.

**Review by ratik21 — COMMENTED** [review]
Indentation does not seem right while viewing file. 
- image
- image

File: examples/stateful-counter/README.md (hunk 17 above)
Comment (amityadav0) [reply examples/stateful-counter/README.md:1]: Could you please suggest something ?

**Review by ratik21 — APPROVED** [review]
Let's rename `scripts/interaction_scripts` to `scripts/interaction` (since we're already inside scripts). 
Otherwise lgtm, tested on my system.

**Pull request merged**